stdlib/public/Concurrency/AsyncPrefixWhileSequence.swift

```diff
@@ -0,0 +1,80 @@
+////===----------------------------------------------------------------------===//
+////
+//// This source file is part of the Swift.org open source project
+////
+//// Copyright (c) 2021 Apple Inc. and the Swift project authors
+//// Licensed under Apache License v2.0 with Runtime Library Exception
+////
+//// See https://swift.org/LICENSE.txt for license information
+//// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+////
+////===----------------------------------------------------------------------===//
+
+import Swift
+
+extension AsyncSequence {
+  @inlinable
+  public __consuming func prefix(
+    while predicate: @escaping (Element) async -> Bool
+  ) rethrows -> AsyncPrefixWhileSequence<Self> {
+    return AsyncPrefixWhileSequence(self, predicate: predicate)
+  }
+}
+
+@frozen
+public struct AsyncPrefixWhileSequence<Upstream: AsyncSequence> {
+  public let upstream: Upstream
+  public let predicate: (Upstream.Element) async -> Bool
+
+  @inlinable
+  public init(
+    _ upstream: Upstream, 
+    predicate: @escaping (Upstream.Element) async -> Bool
+  ) {
+    self.upstream = upstream
+    self.predicate = predicate
+  }
+}
+
+extension AsyncPrefixWhileSequence: AsyncSequence {
+  public typealias Element = Upstream.Element
+  public typealias AsyncIterator = Iterator
+
+  @frozen
+  public struct Iterator: AsyncIteratorProtocol {
+    @usableFromInline
+    var predicateHasFailed = false
+
+    @usableFromInline
+    var upstreamIterator: Upstream.AsyncIterator
+
+    @usableFromInline
+    let predicate: (Upstream.Element) async -> Bool
+
+    @usableFromInline
+    init(
+      _ upstreamIterator: Upstream.AsyncIterator, 
+      predicate: @escaping (Upstream.Element) async -> Bool
+    ) {
+      self.upstreamIterator = upstreamIterator
+      self.predicate = predicate
+    }
+
+    @inlinable
+    public mutating func next() async rethrows -> Upstream.Element? {
+      if !predicateHasFailed, let nextElement = try await upstreamIterator.next() {
+        if await predicate(nextElement) {
+          return nextElement
+        } else {
+          predicateHasFailed = true
```

Comment:
I'm probably missing something, but is storing this state useful or can we just return nil right away here? We're going to return nil after falling out of this block, which will stop enumeration, so next will never be called again anyway.

Reply:
I’m not sure that all of these implementations will consistently follow a rule of continuing to return nil after throws or returning nil. Probably need to double check in testing. I would have guessed it was a programmer error (assert).

Reply:
`Swift.IteratorProtocol` requires that after `next()` returns `nil` once, all subsequent calls return `nil`. We should have the same requirement for async iterators.

`next()` throwing is a new thing, so we _could_ just leave it unspecified, but if we can provide the same guarantee that would be ideal.

Reply:
throwing should be isomorphic to returning nil as a rule.